### ISSUE TITLE
Generate `compile_commands.json`

### ISSUE THREAD

**Issue by foip** [zerodaycode/Zork#78]
## Feature Request: Generate `compile_commands.json`

We already collect the compiler commands in the cache.
If we put them inside a [`compile_commands.json`](https://clang.llvm.org/docs/JSONCompilationDatabase.html) users could link or copy it to their project root, which would [enable all tools in the clang suite](https://clang.llvm.org/docs/HowToSetupToolingForLLVM.html) like clangd or clang-tidy.

With this change users could use a setup with VS Code and the clangd extension for example and work with a functioning language server out of the box.

**Comment by foip** [comment]
> Where do we dump the compile_commands.json? In the out directory?

CMake dumps it directly in the `build` directory.
`compile_commands.json` should definitely live in the `out` directory as it might contain absolute paths.
It should also not be checked into version control.

Clang tools only work with a single `compile_commands.json` so there is not really a reason to generate multiple.

> We must provide documentation about how this can be linked against language servers, for example, a full example where clang tools are able to read the generated commands.

`compile_commands.json` is usually just linked from the `build` directory to the project root

~~~bash
ln -s $PROJECT_ROOT/out/compile_commands.json $PROJECT_ROOT/
~~~

I'm not sure how its usually done on windows because NTFS doen't support that kind of soft links.

> One of the things that I would like to implement, is to enable a switch in the config file to enable conditional compilation.
With this switch enabled, we will read the metadata of the files, and we will compare it's latest modification timestamp with the latest time that Zork runned. So, we will generate commands for modified files after the last run (similar to what Cargo does), potentially avoding large build process for large projects.

If we overwrite `compile_commands.json` with commands of a partial / incremental build then the files that weren't compiled are not considered by the clang tools. So that is something to look out for.

We could assemble the compiler commands once for every source file in the project, cache them and only regenerate the commands of files that have changed.
Just as an idea we could keep the commands in a json that maps the source file path to the generated commands, which we could load as a `HashMap` in Rust. Something like this maybe

~~~json
{
    "modules": {
        "src/ifcs/math.cppm": {
            "precompilation": ["clang++", "--precompile", "..."],
            "pcm_file": "out/.../math.pcm",
            "compilation": ["clang++", "..."],
            "object_file": "out/.../math.o",
        }
    },
    "implementations": { "src/impl/math.cpp": "..." }
}
~~~

This would need some changes to our `Command` and `Argument` approach in Rust, but we would have a nice place where we keep the source files and the artifacts of those source files in the same place.

**Comment by Pyzyryab** [comment]
Could we join the idea of the `HashMap` for the `compile_commands.json` with our cache? Does it makes sense?

`Argument<'a>` could receive a generic parameter of another kind of type:
```
Argument<'a, T> {...} where T: `Kind`
```
or have some kind of enumeratated type to represent all the possible arguments that we could build (precompilation, object_file, compilation...)

**Comment by foip** [comment]
I'm working on this Issue right now. I'll try out some stuff for now and see what works.
The thing with `compile_commands.json` is that it just contains source compilation and module precompilation, so it doesn't contain all commands that we generate (at least that's how other tools that generate `compile_commands.json` handle it).
The link commands and Module compilations (from `.pcm` to `.o`) would be missing.

It just contains a list as specified by the format from [my link above](https://clang.llvm.org/docs/JSONCompilationDatabase.html), so we would have to convert it to a `HashMap` internally.
Its purpose is as input to clang tools, but not as an efficient cache.

**Comment by Pyzyryab** [comment]
Hmmm.... didn't notice about that. Good look, I really want to see the changes

**Comment by Pyzyryab** [comment]
I am taking this feature proposal today, to implementing and merge it with the latest changes in the `0.7.0` release, since I am not seeing activity into upstream for this issue. This doesn't means that, if you have changes, will be ignored/discarded. I just want to try out today that I have free time to generate the `compile_commands.json` and see what happens, following the expected structure by the Clang's official documentation.

**Comment by foip** [comment]
Sounds Great.

Everything I tried so far either didn't work or resulted in rewriting large portions of the project. Also I don't have much time to work on this right now, so I'm excited what you come up with.

**Comment by foip** [comment]
The most promising approach I tried was splitting the build process.

I tried writing an extra function `update_cache` where I tried to determine what needed to be built (files changed). It would generate the according commands and dump them in the cache.

Then the idea was to pass that info and cached commands to the `build` function itself.

Then you would just have to translate the command cache to the `json` format for clang.

Maybe this idea helps.

**Comment by Pyzyryab** [comment]
Sure. Is a nice approach, but I followed other way, just for not change too much the codebase. The main idea is that Rust is extremely fast generating the cache and the compile_commands database, so I am trying to don't worry to much about this (benchmarks don't show neither a 5% of performance lost), and keep the codebase the most stable posible (whenever is possible).

Even tho, I would appreciate if you got some time to check the changes in the code, and propose corrections/upgrades that you consider that are worth the time

**Comment by Pyzyryab** [comment]
closed by #90
